library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java

```diff
@@ -286,271 +146,233 @@ public AbstractJdbcPollInputOperator()
-            bound = previousUpperBound;
-          }
-          ps = store.getConnection().prepareStatement(
-              JdbcMetaDataUtility.buildPollableQuery(getTableName(), getKey(), bound),
-              java.sql.ResultSet.TYPE_FORWARD_ONLY, java.sql.ResultSet.CONCUR_READ_ONLY);
-          isPollable = true;
-        }
-        isReplayed = false;
-        LOG.debug("Prepared statement after re-initialization - {} ", ps.toString());
-      } catch (SQLException e) {
-        // TODO Auto-generated catch block
-        throw new RuntimeException(e);
-      }
+    if (isPollerPartition) {
+      updatePollQuery();
+      isPolled = false;
     }
+    lowerBound = lastEmittedRecord;
+  }
 
-    //Reset the pollable query with the updated upper and lower bounds
-    if (isPollable) {
+  private void updatePollQuery()
+  {
+    if ((lastPolledBound != lastEmittedRecord)) {
+      if (lastEmittedRecord == null) {
+        lastPolledBound = rangeQueryPair.getKey();
+      } else {
+        lastPolledBound = lastEmittedRecord;
+      }
       try {
-        String bound = null;
-        if (previousUpperBound == null && highestPolled == null) {
-          bound = getRangeQueryPair().getKey();
-        } else {
-          bound = highestPolled;
-        }
-        ps = store.getConnection().prepareStatement(
-            JdbcMetaDataUtility.buildPollableQuery(getTableName(), getKey(), bound),
-            java.sql.ResultSet.TYPE_FORWARD_ONLY, java.sql.ResultSet.CONCUR_READ_ONLY);
-        LOG.debug("Polling query {} {}", ps.toString(), currentWindowId);
-        isPolled = false;
+        ps = store.getConnection().prepareStatement(buildRangeQuery(lastPolledBound, Integer.MAX_VALUE),
+            TYPE_FORWARD_ONLY, CONCUR_READ_ONLY);
       } catch (SQLException e) {
         throw new RuntimeException(e);
       }
     }
 
-    lower = null;
-    upper = null;
-
-    //Check if a thread is already active and start only if its no
-    //Do not start the thread from setup, will conflict with the replay
-    if (dbPoller == null && !isReplayed) {
-      //If this is not a replayed state, reset the ps to highest read offset + 1, 
-      //keep the upper bound as the one that was initialized after static partitioning
-      LOG.info("Statement when re-initialized {}", ps.toString());
-      dbPoller = new Thread(new DBPoller());
-      dbPoller.start();
-    }
   }
 
   @Override
   public void emitTuples()
   {
-    if (isReplayed) {
+    if (currentWindowId <= windowManager.getLargestRecoveryWindow()) {
       return;
     }
-
-    List<T> tuples;
-
-    if ((tuples = emitQueue.poll()) != null) {
-      for (Object tuple : tuples) {
-        if (lower == null) {
-          lower = tuple.toString();
-        }
-        upper = tuple.toString();
-        outputPort.emit((T)tuple);
+    int pollSize = (emitQueue.size() < batchSize) ? emitQueue.size() : batchSize;
+    while (pollSize-- > 0) {
+      T obj = emitQueue.poll();
+      if (obj != null) {
```

Comment:
As we are checking queuesize before entering the while loop, there is rare chance that "obj" will be null. I had added the check just for safety.